lucille-core/src/main/java/com/kmwllc/lucille/connector/storageclient/StorageClient.java

```diff
@@ -0,0 +1,108 @@
+ * and publish files to the Lucille pipeline.
+ *
+ *  - create : create appropriate client based on the URI scheme with authentication/settings from cloudOptions
+ *  - init : Initialize the client
+ *  - shutdown : Shutdown the client
+ *  - traverse : traverse through the storage client and publish files to Lucille pipeline
+ *  - validateCloudOptions : Validate that respective cloud authentication information is present for the given cloud provider.
+ *    Does not validate the correctness of the authentication information.
+ */
+
+public interface StorageClient {
+
+  /**
+   * Initialize the client and creates necessary connections and/or resources
+   */
+  void init() throws ConnectorException;
+
+  /**
+   * Shutdown the client, closes any open connections and/or resources
+   */
+  void shutdown() throws IOException;
+
+  /**
+   * Traverses through the storage client and publish files to Lucille pipeline
+   */
+  void traverse(Publisher publisher) throws Exception;
+
+  /**
+   * Gets the appropriate client based on the URI scheme and validate with authentication/settings from cloudOptions
+   */
+  static StorageClient create(URI pathToStorage, String docIdPrefix, List<Pattern> excludes, List<Pattern> includes,
+      Map<String, Object> cloudOptions, Config fileOptions) {
+    String activeClient = pathToStorage.getScheme() != null ? pathToStorage.getScheme() : "file";
+    switch (activeClient) {
+      case "gs" -> {
+        validateCloudOptions(pathToStorage, cloudOptions);
+        return new GoogleStorageClient(pathToStorage, docIdPrefix, excludes, includes, cloudOptions, fileOptions);
+      }
+      case "s3" -> {
+        validateCloudOptions(pathToStorage, cloudOptions);
+        return new S3StorageClient(pathToStorage, docIdPrefix, excludes, includes, cloudOptions, fileOptions);
+      }
+      case "https" -> {
+        String authority = pathToStorage.getAuthority();
+        if (authority != null && authority.contains("blob.core.windows.net")) {
+          validateCloudOptions(pathToStorage, cloudOptions);
+          return new AzureStorageClient(pathToStorage, docIdPrefix, excludes, includes, cloudOptions, fileOptions);
+        } else {
+          throw new IllegalArgumentException("Unsupported client type: " + activeClient + " with authority: " + authority);
+        }
+      }
+      case "file" -> {
+        return new LocalStorageClient(pathToStorage, docIdPrefix, excludes, includes, cloudOptions, fileOptions);
+      }
+      default -> throw new IllegalArgumentException("Unsupported client type: " + activeClient);
+    }
+  }
+
+  /**
+   * Validate that respective cloud authentication information is present for the given cloud provider.
+   * Does not validate the correctness of the authentication information.
+   */
+  static void validateCloudOptions(URI storageURI, Map<String, Object> cloudOptions) {
+    switch (storageURI.getScheme()) {
+      case "gs":
```

Comment:
[nit] I think it would be a little cleaner if each StorageClient had a validateCloudOptions method which validates its own options. Then this method would just have the switch statement to determine which one is called.

Reply:
simplified in [2ee666a](https://github.com/kmwtechnology/lucille/pull/217/commits/2ee666a18534e713f42c9a4ab0c4fabd84b5f477)